### PR TITLE
refactor(chunking): refactor MySQL helpers for unit testing

### LINKED ISSUE #928: Separate chunking-related logic from this function so the individual components can be unit tested independently.
https://github.com/datazip-inc/olake/blob/06aeb9260e7c348c8bd74ed3d93e1ba0023e2e71/drivers/mysql/internal/backfill.go#L64-L69

---

###### This issue was generated by [todo-issue](https://github.com/DerJuulsn/todo-issue) based on a `TODO` comment in 06aeb9260e7c348c8bd74ed3d93e1ba0023e2e71. It&#x27;s been assigned to @saksham-datazip because they committed the code.

### PR DESCRIPTION
**Pull request by vimla01**
# Description

Refactored the MySQL chunking flow in `GetOrSplitChunks` into smaller helper functions so the chunk sizing, chunk column selection, chunk bound selection, and split strategy routing can be unit tested independently.

This keeps the existing behavior the same, but makes the chunking code easier to test and reason about.

Fixes #928

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Ran focused MySQL backfill unit tests:
  `go test ./drivers/mysql/internal -run 'Test(Chunk|LimitOffset|SplitEvenly|PrimaryKey)'`

# Screenshots or Recordings

[olake_928.webm](https://github.com/user-attachments/assets/973466d0-c67f-43a6-838f-132adeafcfba)

## Documentation

- [x] N/A (bug fix, refactor, or test changes only)

## Related PR's (If Any):

N/A

### CODE REVIEW

**Comment by vimla01** [comment]
@nayanj98  could you please take a look at this PR?
I have attached a video showcasing the changes and a successful sync with Olake (iceberg & parquet both)

**Comment by nayanj98** [comment]
@vimla01 Thanks a lot will assign a reviewer for this PR soon

**Comment by nayanj98** [comment]
@vimla01 Assigning @vaibhav-datazip to review your PR

File: drivers/mysql/internal/backfill.go [hunk 1]
```diff
@@ -61,336 +61,363 @@ func (m *MySQL) ChunkIterator(ctx context.Context, stream types.StreamInterface,
...
-		approxRowCount      int64
-		avgRowSize          any
-		approxTableSize     int64
-		columnCollationType string
-		dataMaxLength       sql.NullInt64
-	)
-
-	tableStatsQuery := jdbc.MySQLTableStatsQuery()
-	err := m.client.QueryRowContext(ctx, tableStatsQuery, stream.Name()).Scan(&approxRowCount, &avgRowSize, &approxTableSize)
+	stats, err := m.fetchTableStats(ctx, stream)
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: I don't think we need to abstract out every single logic into a function as these are too basic implementation , primarily we need to abstract out the chunking strategies into separate functions so that those logic be tested.

You can go through mssql driver code once, that has all the chunking strategies separated.

also, you can keep some of the logic in separate functions , after the changes you can add comment on pr description or separate comment, why you think the logic needs a separate function.

File: drivers/mysql/internal/backfill.go [hunk 2]
```diff
@@ -61,336 +61,363 @@ func (m *MySQL) ChunkIterator(ctx context.Context, stream types.StreamInterface,
...
-
-		Final Step:
-		- If sufficient boundaries are not obtained, fallback to primary key-based chunking.
-		- Reduce boundary count to match expectedChunks if needed by evenly subsampling while preserving order.
-
-		Example:
-		minVal = "aa", maxVal = "az", expectedChunks = 4
-
-		Generated boundaries after refining boundaries using collation-aware DB queries:
-		["aa", "ai", "ar", "az"]
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go:288]: this comment is necessary as it explains the necessary, even the other deleted comments should be kept

File: drivers/mysql/internal/backfill.go [hunk 3]
```diff
@@ -61,336 +61,363 @@ func (m *MySQL) ChunkIterator(ctx context.Context, stream types.StreamInterface,
...
-		})
+	if len(chunkBoundaries) < int(math.Ceil(float64(constants.MysqlChunkAcceptanceRatio*float64(expectedChunks)))) {
+		return nil, nil
+	}
+	return chunksFromBoundaries(condenseStrings(chunkBoundaries, expectedChunks)), nil
+}
 
-		logger.Infof("Chunking completed using splitEvenlyForString Method for stream %s", stream.ID())
-		return nil
+func expectedStringChunkCount(approxTableSize int64) int64 {
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: the newly added functions should have a comment on top of them explaining what they do

File: drivers/mysql/internal/backfill.go [hunk 4]
```diff
@@ -478,6 +505,21 @@ func isStringSupportedPK(minVal any, maxVal any, dataMaxLength sql.NullInt64, da
 }
 
 type (
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: move this on top of all the functions

File: drivers/mysql/internal/backfill.go [hunk 5]
```diff
@@ -61,336 +61,363 @@ func (m *MySQL) ChunkIterator(ctx context.Context, stream types.StreamInterface,
...
-	*/
-	splitEvenlyForInt := func(chunks *types.Set[types.Chunk]) error {
+func limitOffsetChunks(approxRowCount, chunkSize int64) *types.Set[types.Chunk] {
+	chunks := types.NewSet[types.Chunk]()
+	chunks.Insert(types.Chunk{
+		Min: nil,
+		Max: utils.ConvertToString(chunkSize),
+	})
+	lastChunk := chunkSize
+	for lastChunk < approxRowCount {
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: the chunking process earlier used to take place in an isolation level , will be better if you go through what was happening earlier and why that has been remove by you now

File: drivers/mysql/internal/backfill_test.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
+package driver
+
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill_test.go:2]: in this file unit tests for some functions are missing , will be better if you can think of edge test cases in those as well as already present functions and add them .

File: drivers/mysql/internal/backfill.go (hunk 1 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go]: Updated this to keep the basic metadata and orchestration logic inside `GetOrSplitChunks`. Only the actual chunking strategies and reusable boundary calculations remain separate

File: drivers/mysql/internal/backfill.go (hunk 2 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go:288]: Restored.

File: drivers/mysql/internal/backfill.go (hunk 3 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go]: added comments above the introduced helper functions .

File: drivers/mysql/internal/backfill_test.go (hunk 6 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill_test.go:2]: Added more focused unit tests for the extracted chunking helpers and existing utilities. I covered edge cases around numeric chunk ranges/overflow, limit-offset boundaries, composite primary key args, string boundary generation, charset encoding, padding, condensing, and strategy eligibility.

File: drivers/mysql/internal/backfill.go (hunk 5 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go]: i checked the previous flow and restored the isolation wrapper for the limit-offset chunking path so the behavior remains the same as before

File: drivers/mysql/internal/backfill.go [hunk 7]
```diff
@@ -132,265 +145,293 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
 		if err != nil {
 			return nil, fmt.Errorf("failed to fetch column datatype and max length for column %s: %s", pkColumns[0], err)
 		}
-		// 1. Try Numeric Strategy
-		numericChunkBounds = isNumericAndEvenDistributed(minVal, maxVal, approxRowCount, chunkSize, dataType)
 
-		// 2. If not numeric, check for supported String strategy
+		// Prefer an arithmetic split for evenly distributed numeric keys.
+		numericChunkBounds = isNumericAndEvenDistributed(minVal, maxVal, approxRowCount, chunkSize, dataType)
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go:151]: can you restore this as well as the comments can be preserved and there dosent seem to be any structural/logical change

File: drivers/mysql/internal/backfill.go [hunk 8]
```diff
@@ -132,265 +145,293 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
-			if minVal == nil {
-				return nil
-			}
-			chunks.Insert(types.Chunk{
-				Min: nil,
-				Max: utils.ConvertToString(minVal),
-			})
+	switch {
+	case numericChunkBounds != nil:
+		logger.Infof("Using splitEvenlyForInt Method for stream %s", stream.ID())
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: can you check other drivers as well for the log placement of chunking strategy being used, and debug being used instead of info , and make it consistent for strategy being used

File: drivers/mysql/internal/backfill.go [hunk 9]
```diff
@@ -132,265 +145,293 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
+		}
+		logger.Infof("Chunking completed using splitEvenlyForString Method for stream %s", stream.ID())
+		return chunks, nil
+	case len(pkColumns) > 0:
+		logger.Infof("Using splitViaPrimaryKey Method for stream %s", stream.ID())
+		return m.splitViaPrimaryKey(ctx, stream, minVal, maxVal, pkColumns, chunkSize)
+	default:
+		logger.Infof("Falling back to limit offset method for stream %s", stream.ID())
+		var chunks *types.Set[types.Chunk]
+		err := jdbc.WithIsolation(ctx, m.client, true, func(_ *sql.Tx) error {
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: shouldn't the isolation level be done inside the function

File: drivers/mysql/internal/backfill.go [hunk 10]
```diff
@@ -132,265 +145,293 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
-		(-∞, "aa"), ["aa","ai"), ["ai","ar"), ["ar","az"), ["az", +∞)
-	*/
-	splitEvenlyForString := func(chunks *types.Set[types.Chunk]) error {
-		expectedChunks := int64(math.Ceil(float64(approxTableSize) / float64(constants.EffectiveParquetSize)))
-		expectedChunks = utils.Ternary(expectedChunks <= 0, int64(1), expectedChunks).(int64)
-
-		stringChunkStepSize := new(big.Int).Sub(stringChunkBounds.maxEncodedBigIntValue, stringChunkBounds.minEncodedBigIntValue)
-		stringChunkStepSize.Add(stringChunkStepSize, new(big.Int).Sub(big.NewInt(expectedChunks), big.NewInt(1)))
-		stringChunkStepSize.Div(stringChunkStepSize, big.NewInt(expectedChunks))
-
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go:300]: even the stringchunkstepsize is being used at one place only , can we use that only instead of defining a function for it

**Comment by vaibhav-datazip** [comment]
@vimla01 once you are done with your changes after the review , please re-request the review from here so that I know the pr is ready again
Screenshot 2026-06-18 at 11 42 03 AM

File: drivers/mysql/internal/backfill.go (hunk 7 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go:151]: Restored.

File: drivers/mysql/internal/backfill.go (hunk 8 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go]: Checked the other drivers as well. Oracle already uses debug for chunking strategy selection, and I updated MySQL and MongoDB strategy-selection logs from Infof to Debugf for consistency.

File: drivers/mysql/internal/backfill.go (hunk 10 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go:300]: Addressed. Removed the separate stringChunkStepSize helper and kept the step-size calculation inline at the single usage site.

File: drivers/mysql/internal/backfill.go (hunk 5 above)
Comment (vaibhav-datazip) [reply drivers/mysql/internal/backfill.go]: I don't think you have restored it everywhere like in limitOffsetChunks

File: drivers/mysql/internal/backfill.go (hunk 7 above)
Comment (vaibhav-datazip) [reply drivers/mysql/internal/backfill.go:151]: the primaryKeyChunkArgs has been abstracted, and in that comments are missing which were present earlier

File: drivers/mysql/internal/backfill.go [hunk 11]
```diff
@@ -132,265 +145,284 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
+	rangeSlice := []string{}
+	adjustedStepSize := new(big.Int).Set(stepSize)
+	adjustedStepSize.Add(adjustedStepSize, big.NewInt(stepShrinkFactor))
+	adjustedStepSize.Div(adjustedStepSize, big.NewInt(stepShrinkFactor+1))
+	currentBoundary := new(big.Int).Set(bounds.minEncodedBigIntValue)
+
+	for chunkIdx := int64(0); chunkIdx < expectedChunks*(stepShrinkFactor+1) && currentBoundary.Cmp(bounds.maxEncodedBigIntValue) < 0; chunkIdx++ {
+		rangeSlice = append(rangeSlice, decodeBigIntToCharsetString(currentBoundary))
+		currentBoundary.Add(currentBoundary, adjustedStepSize)
 	}
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: the test cases are not that necessary and are limited maybe we should put this also back to its original place

File: drivers/mysql/internal/backfill.go [hunk 12]
```diff
@@ -132,265 +145,284 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
 
-	return chunks, err
+	return append(rangeSlice, bounds.MaxPadded)
+}
+
+// chunksFromBoundaries converts ordered boundary values into open-ended chunks.
+func chunksFromBoundaries(boundaries []string) *types.Set[types.Chunk] {
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: similarly for this one

File: drivers/mysql/internal/backfill.go [hunk 13]
```diff
@@ -132,265 +145,284 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
-		(-∞, 0), [0,25), [25,50), [50,75), [75,100), [100, +∞)
-	*/
-	splitEvenlyForInt := func(chunks *types.Set[types.Chunk]) error {
-		chunks.Insert(types.Chunk{
-			Min: nil,
-			Max: utils.ConvertToString(numericChunkBounds.MinBoundary),
-		})
-		prev := numericChunkBounds.MinBoundary
-		for next := numericChunkBounds.MinBoundary + numericChunkBounds.ChunkStep; next <= numericChunkBounds.MaxBoundary; next += numericChunkBounds.ChunkStep {
-			// Guard against int64 overflow causing `next` to wrap and become ≤ prev
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go:244]: comment like these are still pending to be restored , please check again in all the changes of your pr to figure similar things 

you can use GitHub pr extension that might help

File: drivers/mysql/internal/backfill.go [hunk 14]
```diff
@@ -132,265 +145,305 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
-		0 → 25 → 50 → 75 → 100
+// buildLimitOffsetChunks divides a table without a usable key into offset-based chunks.
+func buildLimitOffsetChunks(approxRowCount, chunkSize int64) *types.Set[types.Chunk] {
+	chunks := types.NewSet[types.Chunk]()
+
+	// Start with the first batch-sized range.
+	chunks.Insert(types.Chunk{
+		Min: nil,
+		Max: utils.ConvertToString(chunkSize),
+	})
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: see limit offset chunking itself is a function which should give us the chunks, creating a new function just to separate the chunks and isolation thing dosent make sense
even the test cases for limit offset chunk function wont be any better so we can combine them both again

File: drivers/mysql/internal/backfill.go [hunk 15]
```diff
@@ -132,265 +145,305 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
+		if len(rangeSlice) > len(chunkBoundaries) {
+			chunkBoundaries = slices.Clone(rangeSlice)
+		}
+		if len(rangeSlice) >= int(expectedChunks) {
+			break
+		}
+	}
+
+	if len(chunkBoundaries) < int(math.Ceil(float64(constants.MysqlChunkAcceptanceRatio*float64(expectedChunks)))) {
+		return nil, nil
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: why are not throwing error in this case , as done in int case

File: drivers/mysql/internal/backfill.go [hunk 16]
```diff
@@ -132,265 +145,305 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
-		logger.Infof("Using splitEvenlyForString Method for stream %s", stream.ID())
-		err = splitEvenlyForString(chunks)
-	case len(pkColumns) > 0:
-		logger.Infof("Using splitViaPrimaryKey Method for stream %s", stream.ID())
-		err = splitViaPrimaryKey(stream, chunks)
-	default:
-		logger.Infof("Falling back to limit offset method for stream %s", stream.ID())
-		err = limitOffsetChunking(chunks)
+// expectedStringChunkCount estimates the target number of string-key chunks from table size.
+func expectedStringChunkCount(approxTableSize int64) int64 {
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: even this helper function is simple maths only used once you can put this back

**Review by vaibhav-datazip — COMMENTED** [review]
each time you are done doing the changes , you should ask the reviewer for re review please do the same after making your changes this time as well

File: drivers/mysql/internal/backfill.go (hunk 15 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go]: yes done

File: drivers/mysql/internal/backfill.go (hunk 16 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go]: okay

File: drivers/mysql/internal/backfill.go (hunk 14 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go]: okay  combined limit-offset chunk creation back into limitOffsetChunks

**Comment by vimla01** [comment]
> each time you are done doing the changes , you should ask the reviewer for re review please do the same after making your changes this time as well

sry i missed requesting re-review on the previous one. i had intended to verify the changes the next day first and then request re-review, but got delayed. i'll make sure to do so for future updates.

File: drivers/mysql/internal/backfill.go [hunk 17]
```diff
@@ -132,265 +145,290 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
-			chunks.Insert(types.Chunk{
-				Min: nil,
-				Max: utils.ConvertToString(minVal),
-			})
+	switch {
+	case numericChunkBounds != nil:
+		logger.Debugf("Using splitEvenlyForInt Method for stream %s", stream.ID())
+		chunks, err := splitEvenlyForInt(numericChunkBounds)
+		if err != nil {
+			logger.Warnf("int64 arithmetic overflow, falling back to splitViaPrimaryKey for stream %s", stream.ID())
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: see u are sending the error from the above function but its silently getting eaten , you should print that in this message 

check other places as well where ur changes are silently eating the changes

File: drivers/mysql/internal/backfill.go (hunk 7 above)
Comment (vaibhav-datazip) [reply drivers/mysql/internal/backfill.go:151]: had asked you earlier that you can explain the reason why certain functions are abstracted out [here](https://github.com/datazip-inc/olake/pull/975#discussion_r3408642494), 

if you don't think it was necessary ,can you put primarykeychunkargs function back as well,


A suggestion as well : when we say that some function might not be required to be abstracted out or some other change is suggested. you should check if there are more such changes which follow the similar pattern, and do the suggested change everywhere in the PR.

doing a self review of your pr before you push the changes for re review will help a lot, it will help your pr to get merged much sooner as well and with less comments

File: drivers/mysql/internal/backfill_test.go [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+	"github.com/datazip-inc/olake/types"
+)
+
+func TestSplitEvenlyForInt(t *testing.T) {
+	tests := []struct {
+		name      string
+		bounds    *NumericChunkBounds
+		expected  []types.Chunk
+		wantError bool
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill_test.go]: use the naming convention used in other files , check _test files present in the codebase like reformat_test, flatten_test etc.

File: drivers/mysql/internal/backfill.go [hunk 19]
```diff
@@ -179,7 +179,31 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
 		return m.splitViaPrimaryKey(ctx, stream, minVal, maxVal, pkColumns, chunkSize)
 	default:
 		logger.Debugf("Falling back to limit offset method for stream %s", stream.ID())
-		return m.limitOffsetChunks(ctx, stream, approxRowCount, chunkSize)
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: this is a chunking strategy please keep it separate only like you had done earlier , the comment was only for primaryKeyChunkArgs function, chunking strategies were supposed to be abstracted out so that unit tests can be added for it

File: drivers/mysql/internal/backfill_test.go [hunk 20]
```diff
@@ -135,21 +103,21 @@ func TestIsNumericAndEvenDistributed(t *testing.T) {
 		approxRowCount int64
 		chunkSize      int64
 		dataType       string
-		wantBounds     bool
+		expected       bool
 	}{
-		{name: "supported bigint", minVal: 1, maxVal: 100, approxRowCount: 100, chunkSize: 10, dataType: "BIGINT", wantBounds: true},
+		{name: "supported bigint", minVal: 1, maxVal: 100, approxRowCount: 100, chunkSize: 10, dataType: "BIGINT", expected: true},
 		{name: "unsupported type", minVal: 1, maxVal: 100, approxRowCount: 100, chunkSize: 10, dataType: "varchar"},
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill_test.go]: also please can you add comments for the test cases you have added in this file , for reference you can check other unit test files like resolver_test, reformat_test , etc.

File: drivers/mysql/internal/backfill.go [hunk 21]
```diff
@@ -132,265 +145,287 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
-				// Split the current value into parts
-				columns := strings.Split(utils.ConvertToString(currentVal), ",")
-
-				// Create args array with the correct number of arguments for the query
-				args := make([]interface{}, 0)
-				for columnIndex := 0; columnIndex < len(pkColumns); columnIndex++ {
-					// For each column combination in the WHERE clause, we need to add the necessary parts
-					for partIndex := 0; partIndex <= columnIndex && partIndex < len(columns); partIndex++ {
-						args = append(args, columns[partIndex])
-					}
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go:171]: can please add these comments back

File: drivers/mysql/internal/backfill.go [hunk 22]
```diff
@@ -132,265 +145,287 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
-				}
-				var nextValRaw interface{}
-				err := tx.QueryRowContext(ctx, query, args...).Scan(&nextValRaw)
-				if err == sql.ErrNoRows || nextValRaw == nil {
-					break
-				} else if err != nil {
-					return fmt.Errorf("failed to get next chunk end: %s", err)
-				}
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go:179]: in the new changes you have created two if blocks , is there a reason for that of can we keep this if and else if structure only

File: drivers/mysql/internal/backfill.go [hunk 23]
```diff
@@ -132,265 +145,287 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
-				args := make([]interface{}, 0)
-				for columnIndex := 0; columnIndex < len(pkColumns); columnIndex++ {
-					// For each column combination in the WHERE clause, we need to add the necessary parts
-					for partIndex := 0; partIndex <= columnIndex && partIndex < len(columns); partIndex++ {
-						args = append(args, columns[partIndex])
-					}
-				}
-				var nextValRaw interface{}
-				err := tx.QueryRowContext(ctx, query, args...).Scan(&nextValRaw)
-				if err == sql.ErrNoRows || nextValRaw == nil {
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go:175]: earlier nextval raw nil was also getting checked in this condition only , any reason you have made it a separate condition or can that be reverted back to this only ?

File: drivers/mysql/internal/backfill.go [hunk 24]
```diff
@@ -132,265 +145,287 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
+["aa", "ai", "ar", "az"]
+
+Chunks:
+(-inf, "aa"), ["aa","ai"), ["ai","ar"), ["ar","az"), ["az", +inf)
+*/
+func (m *MySQL) splitEvenlyForString(ctx context.Context, stream types.StreamInterface, bounds *StringChunkBounds, pkColumn, columnCollationType string, approxTableSize int64) (*types.Set[types.Chunk], error) {
+	expectedChunks := int64(math.Ceil(float64(approxTableSize) / float64(constants.EffectiveParquetSize)))
+	if expectedChunks <= 0 {
+		expectedChunks = 1
 	}
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: can you revert this back to the utils.ternary condition only

File: drivers/mysql/internal/backfill.go [hunk 25]
```diff
@@ -132,265 +145,287 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
+			currentBoundary.Add(currentBoundary, adjustedStepSize)
+		}
+
+		rangeSlice = append(rangeSlice, bounds.MaxPadded)
+
+		// Align boundaries with actual DB values using MySQL collation ordering
+		query, args := jdbc.MySQLDistinctAlignedPKValuesWithCollationQuery(stream, pkColumn, rangeSlice, columnCollationType, bounds.MinPadded, bounds.MaxPadded)
+		rows, err := m.client.QueryContext(ctx, query, args...)
+		if err != nil {
+			boundaryQueryErr = fmt.Errorf("failed to query distinct string chunk boundaries: %w", err)
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: in this code base we are using %s for errors instead of %w , so please revert %w to %s wherever you have done changes, you can check the other files of codebase as well to verify this , 

also can you use the older message for distinct boundary error

distinct boundary query failed for stream %s: %s

File: drivers/mysql/internal/backfill.go (hunk 22 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go:179]: I had separated the checks while reviewing error handling, but the earlier if/else-if structure was better aligned with the existing flow. ill keep that

File: drivers/mysql/internal/backfill.go (hunk 23 above)
Comment (vimla01) [reply drivers/mysql/internal/backfill.go:175]: Reverted it back to checking `sql.ErrNoRows || nextValRaw == nil` in the same condition with the existing `else if err != nil` block.

File: drivers/mysql/internal/backfill.go [hunk 26]
```diff
@@ -132,265 +145,285 @@ func (m *MySQL) GetOrSplitChunks(ctx context.Context, pool *destination.WriterPo
...
+		rangeSlice = rangeSlice[:0]
+		for rows.Next() {
+			var val string
+			if err := rows.Scan(&val); err != nil {
+				rows.Close()
+				return nil, fmt.Errorf("failed to scan row: %s", err)
 			}
+			rangeSlice = append(rangeSlice, val)
 		}
+		rows.Close()
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill.go]: I don't think this will be required, once you can read comments on Close as well , it says rows.next will automatically close it once finished . 

so this rows.Close can be removed

File: drivers/mysql/internal/backfill_test.go [hunk 27]
```diff
@@ -0,0 +1,257 @@
...
+import (
+	"database/sql"
+	"math"
+	"reflect"
+	"testing"
+
+	"github.com/datazip-inc/olake/types"
+)
+
+func TestSplitEvenlyForInt(t *testing.T) {
```
Comment (vaibhav-datazip) [inline drivers/mysql/internal/backfill_test.go:12]: your unit tests are not running on the CI , you can check it in the passed workflows on your PR, it wont list any of the tests added there.

Some change might be required in unit-tests.yml file, can you check it ? also make it generic so that it works for all the files added in future as well

Comment (vimla01) [reply drivers/mysql/internal/backfill_test.go:12]: Got it, I'll look into it.

File: .github/workflows/unit-tests.yml [hunk 28]
```diff
@@ -19,4 +19,6 @@ jobs:
         run: go mod download
 
       - name: Run All Unit Tests
-        run: go test -v ./...
+        run: |
+          TEST_PACKAGES=$(git ls-files '*_test.go' | xargs -n1 dirname | sort -u | sed 's#^#./#' | xargs go list -f '{{ .ImportPath }}' | grep -v '/drivers/db2/internal$')
+          go test -v -count=1 -skip 'Integration|2PC|Performance' $TEST_PACKAGES
```
Comment (vaibhav-datazip) [inline .github/workflows/unit-tests.yml]: can you check if this will be the better approach , as it will help in running db2 unit tests as well

```suggestion
      - name: Install DB2 CLI Driver
        run: |
          go run -C drivers/db2 github.com/ibmdb/go_ibm_db/installer@v0.4.5 "$(pwd)"
          IBM_DB_HOME="$(pwd)/clidriver"
          echo "IBM_DB_HOME=$IBM_DB_HOME" >> $GITHUB_ENV
          echo "CGO_CFLAGS=-I$IBM_DB_HOME/include" >> $GITHUB_ENV
          echo "CGO_LDFLAGS=-L$IBM_DB_HOME/lib -Wl,-rpath,$IBM_DB_HOME/lib" >> $GITHUB_ENV
          echo "LD_LIBRARY_PATH=$IBM_DB_HOME/lib" >> $GITHUB_ENV

      - name: Run All Unit Tests
        run: |
          go list -m | awk '{print $1"/..."}' | xargs go test -v -count=1 -skip 'Integration|2PC|Performance'

```

Comment (vaibhav-datazip) [reply .github/workflows/unit-tests.yml]: you can check the db2 driver installation yml from integration test yml file as well

Comment (vimla01) [reply .github/workflows/unit-tests.yml]: yeah, i ve made the suggested changes. Thanks!!

**Review by vaibhav-datazip — APPROVED** [review]
LGTM
